### DISCUSSION TITLE
Difference about how different objects are created

### DISCUSSION THREAD

**Discussion by DocSepp** [bacnet-stack/bacnet-stack#591] [General]
Hey,

I have a question related to the way different objects are generated in the library, or more specifically why it differs.

If you take the analog inputs for example, when the init function is called, we automatically create as many objects as specified by `MAX_ANALOG_INPUTS`, whereas with analog outputs we only create a keylist using `Keylist_Create()` and are then able to create a (seemingly) arbitrary number of objects dynamically.

Is this somehow required by the standard or is there any other reason?

Thanks in advance!

**Comment by skarg** [comment]
BACnet includes CreateObject and DeleteObject services - both optional - and having dynamic objects enables those services to be used for those devices that want that feature.  However, even for embedded microcontrollers, the ability to create zero, one or more (unlimited) objects according to the needs (object model) of the device simplifies software development and maintenance.  It also shifts the RAM usage to HEAP.

The basic objects (demo, examples) started as fixed size array of objects to be examples for the apps/server (or in the ports for microcontrollers) to showcase the BACnet objects. In several products I developed, I used the keylist library in the BACnet object code to create dynamic BACnet objects - typically at startup - and I have a collection of those library objects that I had planned to bring into the bacnet/basic/object library. Using dynamically creatable objects that can be used both microcontrollers and large devices makes the software maintenance of the example objects easier as I can slim down the number of duplicate example objects in the library.

I'm also in the process of refactoring the device object into the service handlers (bacnet/basic/services/h_device.c) to remove the library dependency on the device.c/device.h files.

**Reply by DocSepp** [reply]
Hello,

thank you for your quick response. I didn't necessarily think about the CreateObject services but it does make sense.
I had the same thought that the code for the different objects could be streamlined if all of them used dynamic allocation of memory.

So as for the objects in the bacnet/basic/object library (I'm mostly talking about analog and binary inputs/outputs as these are the objects my client wants in their product for now), you are planning to make them dynamically creatable, did I understand that correctly?

**Reply by skarg** [reply]
Yes, I am planning to make all the example bacnet/basic/object/ dynamically creatable.

Some of the existing object examples include event and read-range functionality, and rather than remove those features (by copying my object files directly), I'll just port the dynamic features into those examples.
